actix-multipart-derive/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+                rust_name,
+                limit,
+                path,
+            })
+        })
+        .collect::<Result<Vec<_>, darling::Error>>()
+    {
+        Ok(attrs) => attrs,
+        Err(e) => return e.write_errors().into(),
+    };
+
+    // Check that field names are unique
+    let mut set = HashSet::new();
+    for f in &parsed {
+        if !set.insert(f.serialization_name.clone()) {
+            panic!("Multiple fields named: `{}`", f.serialization_name);
+        }
+    }
+
+    // Return value when a field name is not supported by the form
+    let unknown_field_result = if attrs.deny_unknown_fields {
+        quote!(Err(::actix_multipart::MultipartError::UnsupportedField(
+            field.name().to_string()
+        )))
+    } else {
+        quote!(Ok(()))
+    };
+
+    // Value for duplicate action
+    let duplicate_action = match attrs.duplicate_action {
+        DuplicateAction::Ignore => quote!(::actix_multipart::form::DuplicateAction::Ignore),
+        DuplicateAction::Deny => quote!(::actix_multipart::form::DuplicateAction::Deny),
+        DuplicateAction::Replace => quote!(::actix_multipart::form::DuplicateAction::Replace),
+    };
+
+    // read_field() implementation
+    let mut read_field_impl = quote!();
+    for field in &parsed {
+        let name = &field.serialization_name;
+        let path = &field.path;
+        read_field_impl.extend(quote!(
+            #name => #path::handle_field(req, field, limits, state, #duplicate_action).boxed_local(),
+        ));
+    }
+
+    // limit() implementation
+    let mut limit_impl = quote!();
+    for field in &parsed {
+        let name = &field.serialization_name;
+        if let Some(value) = field.limit {
+            limit_impl.extend(quote!(
+                #name => Some(#value),
+            ));
+        }
+    }
+
+    // from_state() implementation
+    let mut from_state_impl = quote!();
+    for field in &parsed {
+        let name = &field.serialization_name;
+        let rust_name = &field.rust_name;
+        let path = &field.path;
+
+        from_state_impl.extend(quote!(
+            #rust_name: #path::from_state(#name, &mut state)?,
+        ));
+    }
+
+    let gen = quote! {
+        impl ::actix_multipart::form::MultipartFormTrait for #name {
+            fn limit(field_name: &str) -> Option<usize> {
+                match field_name {
+                    #limit_impl
+                    _ => None,
+                }
+            }
+
+            fn handle_field<'t>(
+                req: &'t ::actix_web::HttpRequest,
+                field: ::actix_multipart::Field,
+                limits: &'t mut ::actix_multipart::form::Limits,
+                state: &'t mut ::actix_multipart::form::State,
+            ) -> ::futures_core::future::LocalBoxFuture<'t, Result<(), ::actix_multipart::MultipartError>> {
+                use ::actix_multipart::form::FieldGroupReader;
+                use ::futures_util::{TryFutureExt, FutureExt};
```

Comment:
This is a bug I need to fix - the macro implementation requires some dependencies from futures_util and futures_core - so if the end user's crate didn't declare them as dependencies it would currently fail to compile when using the macro.

Should I re-export these (maybe as doc-hidden?) from actix-multipart so that they may then be reliably used from within the macro?

Reply:
reexporting seems best, that's what I've seen other libraries do in this case

it also doesn't look like you need `TryFutureExt` from what I can see, and the others (`LocalBoxFuture` and `boxed_local`) are basically one liners. if you don't anticipate using more deps from futures, it'd be trivial to reimplement

Reply:
Thanks, yep you are right, these are very simple to just write directly